fri/src/prover.rs

```diff
@@ -104,13 +112,27 @@ where
         }
     }
 
-    // We should be left with `blowup` evaluations of a constant polynomial.
-    assert_eq!(folded.len(), config.blowup());
-    let final_poly = folded[0];
-    for x in folded {
-        assert_eq!(x, final_poly);
+    // Previously, we checked that folded was constant at this point. Now, we transform folded
+    // from the evaluation domain to the coefficient domain and verify the trailing coefficients.
+    reverse_slice_index_bits(&mut folded);
+    // TODO: For better performance, we could run the IDFT on only the first half
+    //       (or less, depending on `log_blowup`) of `final_poly`.
+    let final_poly = debug_span!("idft final poly").in_scope(|| Radix2Dit::default().idft(folded));
```

Comment:
I think there is a mild subtlety here related to domains that it would be worth commenting on.

While our trace starts as evaluations over a subgroup `H`, by the time we get to applying `Fri` we have
done an `lde` along with the `log_blow_up` meaning what we have is evaluations over a coset `gK` where `K` is some larger subgroup.

Then, in each `fri` folding step, we replace the domain `gK` with `(g^2)(K^2)` where `K^2` is the unique subgroup of half the size. Hence after applying a large number of folding steps, our final set of evaluations will be over the coset `hJ` where `h = g^{2^n}` and `J = K^{2^n}`.

On the other hand, the standard `Dft`, on a list of size `|J|` passes between the evaluations and coefficient basis for polynomials over the subgroup `J`.

This is a slight domain mismatch! This means that there are `2` different polynomials here. `F(x)` is the polynomial whose evaluations over `hJ` agree with our set of evaluations. `G(x)` is the polynomial whose evaluations over `J` agree with our set of evaluations. The "true" polynomial after out `fri` folding steps should be `F(x)` but our reported polynomial here is `G(x)`.

Luckily for us, in this situation `F(x)` and `G(x)` are closely related by `G(x) = F(hx)` and so the degrees of these two polynomials are the same.

However this could break if we made minor changes elsewhere in the code. In particular if, instead of doing a `log_blow_up` onto a coset `gK`, we did it onto a union of cosets `g_0 H u g_1 H u ... g_n H`. (Is this something we might do @dlubarov ?)

Reply:
Comments added. Thanks for the reviews!